src/webgpu/api/validation/render_pipeline/inter_stage.spec.ts

```diff
@@ -1,10 +1,337 @@
+      { output: '@interpolate(perspective, center) ', input: '@interpolate(perspective, sample) ' },
+      {
+        output: '@interpolate(perspective, center) ',
+        input: '@interpolate(perspective, centroid) ',
+      },
+      { output: '@interpolate(perspective, centroid) ', input: '@interpolate(perspective) ' },
+    ])
+  )
+  .fn(async t => {
+    const { isAsync, output, input, _success } = t.params;
+
+    const descriptor = t.getDescriptorWithStates(
+      t.getVertexStateWithOutputs([`@location(0) ${output}vout0: f32`]),
+      t.getFragmentStateWithInputs([`@location(0) ${input}fin0: f32`])
+    );
+
+    t.doCreateRenderPipelineTest(isAsync, _success ?? output === input, descriptor);
+  });
+
+g.test('max_shader_variable_location')
+  .desc(
+    `Tests that validation should fail when there is location of user-defined output/input variable >= device.limits.maxInterStageShaderVariables`
+  )
+  .params(u =>
+    u
+      .combine('isAsync', [false, true])
+      // user defined variable location = maxInterStageShaderVariables + locationDelta
+      .combine('locationDelta', [0, -1])
+  )
+  .fn(async t => {
+    const { isAsync, locationDelta } = t.params;
+    const location = t.device.limits.maxInterStageShaderVariables + locationDelta;
+
+    const descriptor = t.getDescriptorWithStates(
+      t.getVertexStateWithOutputs([`@location(${location}) vout0: f32`]),
+      t.getFragmentStateWithInputs([`@location(${location}) fin0: f32`])
+    );
+
+    t.doCreateRenderPipelineTest(
+      isAsync,
+      location < t.device.limits.maxInterStageShaderVariables,
+      descriptor
+    );
+  });
+
+g.test('max_components_count,output')
+  .desc(
+    `Tests that validation should fail when scalar components of all user-defined outputs > max vertex shader output components.`
+  )
+  .params(u =>
+    u.combine('isAsync', [false, true]).combineWithParams([
+      // Number of user-defined output scalar components in test shader = device.limits.maxInterStageShaderComponents + numScalarDelta - 4. (builtin position output counts for space as well)
+      { numScalarDelta: 0, topology: 'triangle-list', _success: true },
+      { numScalarDelta: 1, topology: 'triangle-list', _success: false },
+      { numScalarDelta: 0, topology: 'point-list', _success: false },
+      { numScalarDelta: -1, topology: 'point-list', _success: true },
+    ] as const)
+  )
+  .fn(async t => {
+    const { isAsync, numScalarDelta, topology, _success } = t.params;
+
+    // -4 is for scalars used by @builtin(position)
+    const numScalarComponents = t.device.limits.maxInterStageShaderComponents + numScalarDelta - 4;
+
+    const numVec4 = Math.floor(numScalarComponents / 4);
+    const numTrialingScalars = numScalarComponents % 4;
+    const numUserDefinedInterStageVariables = numTrialingScalars === 0 ? numVec4 + 1 : numVec4;
+    // [1]
+    assert(numUserDefinedInterStageVariables < 16);
```

Comment:
Since we've verified the tests themselves are correct and `max_shader_variable_location` and `max_components_count` are failing anyway (current implementation not taking account for spaces used by `@builtin(position)`). I used `t.device.limits.maxInterStageShaderVariables` and removed the TODOs. So that the test themselves are sealed and let the browser for implementations/suppressions.